File: src/data/courses/dsa/dynamic-programming/grid-matrix-dp/unique-paths-ii/code/python/unique_paths_ii.py
class Solution:
    def uniquePathsWithObstacles(self, grid) -> int:
        n = len(grid[0]); dp = [0]*n; dp[0] = 1
        for row in grid:
            for j in range(n):
                if row[j] == 1: dp[j] = 0
                elif j > 0: dp[j] += dp[j-1]
        return dp[-1]

File: src/data/courses/dsa/dynamic-programming/grid-matrix-dp/unique-paths-ii/code/cpp/unique_paths_ii.cpp
#include <vector>
using namespace std;
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& grid) {
        int n = grid[0].size();
        vector<long> dp(n, 0); dp[0] = 1;
        for (auto& row : grid) {
            for (int j = 0; j < n; j++) {
                if (row[j] == 1) dp[j] = 0;
                else if (j > 0) dp[j] += dp[j-1];
            }
        }
        return dp[n-1];
    }
};